File: KeeperLegacyTests/StoryTests.swift
import XCTest
@testable import KeeperLegacy

// MARK: - Story Tests
// Tests for story event triggers, NPC relationship tracking, act advancement,
// and feature unlocking through story events.

@MainActor
final class StoryTests: XCTestCase {

    // MARK: Story Event Structure

    func testAllEventsHaveUniqueIDs() {
        let ids = StoryEvent.allEvents.map { $0.id }
        let unique = Set(ids)
        XCTAssertEqual(ids.count, unique.count, "All story event IDs must be unique")
    }

    func testAllEventsHaveNonEmptyContent() {
        for event in StoryEvent.allEvents {
            XCTAssertFalse(event.title.isEmpty, "\(event.id) must have a title")
            XCTAssertFalse(event.body.isEmpty,  "\(event.id) must have body text")
            XCTAssertFalse(event.npcID.isEmpty, "\(event.id) must have a npcID")
        }
    }

    func testAllEventNPCIDsReferenceRealNPCs() {
        let validNPCIDs = Set(NPC.mainCast.map { $0.id })
        for event in StoryEvent.allEvents {
            XCTAssertTrue(validNPCIDs.contains(event.npcID),
                          "Event '\(event.id)' references unknown NPC '\(event.npcID)'")
        }
    }

    func testEventsAreInChronologicalActOrder() {
        // No Act II event should appear before an Act I event in the array
        var highestActSeen = 1
        for event in StoryEvent.allEvents {
            XCTAssertGreaterThanOrEqual(event.act.rawValue, highestActSeen,
                "Events should be roughly sorted by act (found Act \(event.act.rawValue) after Act \(highestActSeen))")
            highestActSeen = max(highestActSeen, event.act.rawValue)
        }
    }

    func testActAdvancingEventsUnlockFeatures() {
        // Every event that advances to the next act should also unlock a feature
        // (based on current design — this documents the design contract)
        let capstoneEvents = StoryEvent.allEvents.filter { $0.advancesToNextAct }
        XCTAssertFalse(capstoneEvents.isEmpty, "There should be at least one capstone event")
        for event in capstoneEvents {
            XCTAssertNotNil(event.unlocksFeature,
                            "Capstone event '\(event.id)' should unlock a feature")
        }
    }

    // MARK: PlayerStoryState

    func testInitialStateHasNoCompletedEvents() {
        let state = PlayerStoryState()
        XCTAssertTrue(state.completedEventIDs.isEmpty)
        XCTAssertEqual(state.currentAct, 1)
    }

    func testCompleteEventAddsToCompletedSet() {
        var state = PlayerStoryState()
        state.completeEvent("act1_first_egg")
        XCTAssertTrue(state.hasCompletedEvent("act1_first_egg"))
        XCTAssertFalse(state.hasCompletedEvent("act1_shop_secret"))
    }

    func testCompleteEventIsIdempotent() {
        var state = PlayerStoryState()
        state.completeEvent("act1_first_egg")
        state.completeEvent("act1_first_egg")
        XCTAssertEqual(state.completedEventIDs.count, 1)
    }

    func testRelationshipIncreaseClampsAt100() {
        var state = PlayerStoryState()
        state.increaseRelationship(npcID: "elder_mira", by: 95)
        state.increaseRelationship(npcID: "elder_mira", by: 95)
        let level = state.npcRelationshipLevels["elder_mira"] ?? 0
        XCTAssertEqual(level, 100, "Relationship should clamp at 100")
    }

    func testInitialRelationshipLevelsMatchNPCDefaults() {
        let state = PlayerStoryState()
        // Elder Mira starts at 50 per NPC.mainCast definition
        XCTAssertEqual(state.npcRelationshipLevels["elder_mira"], 50)
        // Rival Cass starts at 0
        XCTAssertEqual(state.npcRelationshipLevels["rival_cass"], 0)
    }

    // MARK: StoryViewModel — Trigger Logic

    func testFirstEggTriggersAtLevel1() {
        let vm = StoryViewModel()
        vm.checkTriggers(level: 1)
        XCTAssertNotNil(vm.pendingStoryEvent)
        XCTAssertEqual(vm.pendingStoryEvent?.id, "act1_first_egg")
    }

    func testNoEventTriggersBeforeLevel1() {
        let vm = StoryViewModel()
        vm.checkTriggers(level: 0)
        XCTAssertNil(vm.pendingStoryEvent, "No events should trigger at level 0")
    }

    func testOnlyOneEventPendsAtATime() {
        let vm = StoryViewModel()
        vm.checkTriggers(level: 15)  // Multiple level-based events are eligible
        XCTAssertNotNil(vm.pendingStoryEvent)
        // Only the first eligible unfired event should queue
        XCTAssertEqual(vm.pendingStoryEvent?.id, "act1_first_egg",
                       "Should queue events in allEvents order")
    }

    func testShopSecretTriggersAtLevel10() {
        let vm = StoryViewModel()
        // Pre-complete the first event so act1_shop_secret is next
        vm.storyState.completeEvent("act1_first_egg")
        vm.checkTriggers(level: 10)
        XCTAssertEqual(vm.pendingStoryEvent?.id, "act1_shop_secret")
    }

    func testMagicDiscoveredRequiresPrerequisite() {
        let vm = StoryViewModel()
        vm.storyState.completeEvent("act1_first_egg")
        // act1_shop_secret NOT completed
        vm.checkTriggers(level: 15)
        // act1_shop_secret should fire (level 10 reached), NOT act1_magic_discovered
        XCTAssertEqual(vm.pendingStoryEvent?.id, "act1_shop_secret")
    }

    func testMagicDiscoveredTriggersAfterPrerequisiteCompletes() {
        let vm = StoryViewModel()
        vm.storyState.completeEvent("act1_first_egg")
        vm.storyState.completeEvent("act1_shop_secret")
        vm.checkTriggers(level: 10)
        XCTAssertEqual(vm.pendingStoryEvent?.id, "act1_magic_discovered")
    }

    // MARK: StoryViewModel — Event Completion

    func testCompleteEventUnlocksFeature() {
        let vm         = StoryViewModel()
        let progressVM = ProgressionViewModel()
        progressVM.coins = 10_000

        // Set up: complete prerequisites
        vm.storyState.completeEvent("act1_first_egg")
        vm.storyState.completeEvent("act1_shop_secret")
        vm.pendingStoryEvent = StoryEvent.allEvents.first { $0.id == "act1_magic_discovered" }

        vm.completeCurrentEvent(progressVM: progressVM)

        XCTAssertTrue(progressVM.unlockedFeatures.contains(GameFeature.breeding.rawValue),
                      "Completing act1_magic_discovered should unlock breeding")
    }

    func testCompleteCapstoneEventAdvancesStoryAct() {
        let vm         = StoryViewModel()
        let progressVM = ProgressionViewModel()

        vm.storyState.completeEvent("act1_first_egg")
        vm.storyState.completeEvent("act1_shop_secret")
        vm.pendingStoryEvent = StoryEvent.allEvents.first { $0.id == "act1_magic_discovered" }

        XCTAssertEqual(vm.storyState.currentAct, 1)
        vm.completeCurrentEvent(progressVM: progressVM)
        XCTAssertEqual(vm.storyState.currentAct, 2, "Completing capstone event should advance to Act II")
        XCTAssertEqual(progressVM.storyAct, 2)
    }

    func testCompleteEventBoostsNPCRelationship() {
        let vm         = StoryViewModel()
        let progressVM = ProgressionViewModel()
        let initialLevel = vm.storyState.npcRelationshipLevels["elder_mira"] ?? 0

        vm.pendingStoryEvent = StoryEvent.allEvents.first { $0.id == "act1_first_egg" }
        vm.completeCurrentEvent(progressVM: progressVM)

        let newLevel = vm.storyState.npcRelationshipLevels["elder_mira"] ?? 0
        XCTAssertEqual(newLevel, initialLevel + 10,
                       "Completing an event should boost the delivering NPC's relationship by 10")
    }

    func testCompleteEventClearsPending() {
        let vm         = StoryViewModel()
        let progressVM = ProgressionViewModel()
        vm.pendingStoryEvent = StoryEvent.allEvents[0]
        vm.completeCurrentEvent(progressVM: progressVM)
        XCTAssertNil(vm.pendingStoryEvent)
    }

    func testCompletingEventImmediatelyChecksNext() {
        let vm         = StoryViewModel()
        let progressVM = ProgressionViewModel()
        progressVM.level = 10  // Make act1_shop_secret eligible

        // Complete first event — should immediately queue act1_shop_secret
        vm.pendingStoryEvent = StoryEvent.allEvents.first { $0.id == "act1_first_egg" }
        vm.completeCurrentEvent(progressVM: progressVM)

        XCTAssertEqual(vm.pendingStoryEvent?.id, "act1_shop_secret",
                       "After completing act1_first_egg at level 10, act1_shop_secret should auto-queue")
    }

    // MARK: Story Persistence (DataManager)

    func testSaveAndLoadStoryStateRoundTrips() {
        let dm = DataManager(inMemory: true)
        _ = dm.playerState()

        var state = PlayerStoryState()
        state.completeEvent("act1_first_egg")
        state.completeEvent("act1_shop_secret")
        state.currentAct = 2
        state.increaseRelationship(npcID: "elder_mira", by: 20)

        dm.saveStoryState(state)
        let loaded = dm.loadStoryState()

        XCTAssertTrue(loaded.hasCompletedEvent("act1_first_egg"))
        XCTAssertTrue(loaded.hasCompletedEvent("act1_shop_secret"))
        XCTAssertEqual(loaded.currentAct, 2)
        XCTAssertEqual(loaded.npcRelationshipLevels["elder_mira"], (50 + 20))
    }

    func testFreshLoadReturnsDefaultState() {
        let dm    = DataManager(inMemory: true)
        _ = dm.playerState()
        let state = dm.loadStoryState()

        XCTAssertTrue(state.completedEventIDs.isEmpty)
        XCTAssertEqual(state.currentAct, 1)
    }
}

File: KeeperLegacyGodot/Tests/StoryTests.cs
// Tests/StoryTests.cs
// Port of StoryTests.swift
// Tests story event structure, PlayerStoryState, StoryEngine trigger logic,
// event completion side-effects, and save/load round-trip.

using System;
using System.Collections.Generic;
using System.Linq;
using NUnit.Framework;
using KeeperLegacy.Data;
using KeeperLegacy.Models;

namespace KeeperLegacy.Tests
{
    [TestFixture]
    public class StoryTests
    {
        // ── Event Structure ───────────────────────────────────────────────────

        [Test]
        public void AllEventsHaveUniqueIds()
        {
            var ids    = StoryEvent.AllEvents.Select(e => e.Id).ToList();
            var unique = new HashSet<string>(ids);
            Assert.That(ids.Count, Is.EqualTo(unique.Count), "Duplicate story event IDs found");
        }

        [Test]
        public void AllEventsHaveNonEmptyTitle()
        {
            foreach (var e in StoryEvent.AllEvents)
                Assert.That(e.Title, Is.Not.Empty, $"Event '{e.Id}' has empty title");
        }

        [Test]
        public void AllEventsHaveNonEmptyBody()
        {
            foreach (var e in StoryEvent.AllEvents)
                Assert.That(e.Body, Is.Not.Empty, $"Event '{e.Id}' has empty body");
        }

        [Test]
        public void AllEventsReferenceValidNpcs()
        {
            var npcIds = new HashSet<string>(NPC.MainCast.Select(n => n.Id));
            foreach (var e in StoryEvent.AllEvents)
                Assert.That(npcIds, Does.Contain(e.NpcId),
                    $"Event '{e.Id}' references unknown NPC '{e.NpcId}'");
        }

        [Test]
        public void EventsAreInChronologicalActOrder()
        {
            // Acts should only ever stay the same or increase as we iterate AllEvents
            int maxActSeen = 0;
            foreach (var e in StoryEvent.AllEvents)
            {
                int act = (int)e.Act;
                Assert.That(act, Is.GreaterThanOrEqualTo(maxActSeen),
                    $"Event '{e.Id}' (Act {act}) appears after Act {maxActSeen} — wrong order");
                maxActSeen = act;
            }
        }

        [Test]
        public void CapstoneEventsUnlockFeaturesOrAdvanceAct()
        {
            // Every event marked AdvancesToNextAct must also unlock a feature
            // (Capstones always do both — the act gate is tied to the feature unlock)
            foreach (var e in StoryEvent.AllEvents.Where(ev => ev.AdvancesToNextAct))
                Assert.That(e.UnlocksFeature, Is.Not.Null,
                    $"Capstone event '{e.Id}' advances act but unlocks no feature");
        }

        // ── PlayerStoryState ──────────────────────────────────────────────────

        [Test]
        public void InitialStoryStateIsAct1()
        {
            var state = new PlayerStoryState();
            Assert.That(state.CurrentAct, Is.EqualTo(1));
        }

        [Test]
        public void NoEventsCompletedInitially()
        {
            var state = new PlayerStoryState();
            Assert.That(state.CompletedEventIds, Is.Empty);
        }

        [Test]
        public void CompleteEventAddsToCompletedSet()
        {
            var state = new PlayerStoryState();
            state.CompleteEvent("act1_first_egg");
            Assert.That(state.HasCompletedEvent("act1_first_egg"), Is.True);
        }

        [Test]
        public void CompleteEventIsIdempotent()
        {
            var state = new PlayerStoryState();
            state.CompleteEvent("act1_first_egg");
            state.CompleteEvent("act1_first_egg");
            int count = state.CompletedEventIds.Count(id => id == "act1_first_egg");
            Assert.That(count, Is.EqualTo(1));
        }

        [Test]
        public void HasCompletedEventReturnsFalseForUnknown()
        {
            var state = new PlayerStoryState();
            Assert.That(state.HasCompletedEvent("nonexistent_event"), Is.False);
        }

        [Test]
        public void IncreaseRelationshipAddsAmount()
        {
            var state = new PlayerStoryState();
            state.IncreaseRelationship("elder_mira", 10);
            int level = state.NpcRelationshipLevels["elder_mira"];
            Assert.That(level, Is.GreaterThan(0));
        }

        [Test]
        public void RelationshipClampedAt100()
        {
            var state = new PlayerStoryState();
            state.IncreaseRelationship("elder_mira", 999);
            Assert.That(state.NpcRelationshipLevels["elder_mira"], Is.LessThanOrEqualTo(100));
        }

        [Test]
        public void InitialNpcRelationshipLevelsMatchMainCast()
        {
            var state  = new PlayerStoryState();
            var npcIds = new HashSet<string>(NPC.MainCast.Select(n => n.Id));
            foreach (var id in npcIds)
                Assert.That(state.NpcRelationshipLevels, Does.ContainKey(id),
                    $"Initial state missing relationship for NPC '{id}'");
        }

        // ── StoryEngine Triggers ──────────────────────────────────────────────

        [Test]
        public void NoEventPendingBeforeTriggerCheck()
        {
            var state  = new PlayerStoryState();
            var engine = new StoryEngine(state);
            Assert.That(engine.PendingEvent, Is.Null);
        }

        [Test]
        public void FirstEggEventTriggersAtLevel1()
        {
            var state  = new PlayerStoryState();
            var engine = new StoryEngine(state);
            engine.CheckTriggers(level: 1);
            Assert.That(engine.PendingEvent, Is.Not.Null);
            Assert.That(engine.PendingEvent!.Id, Is.EqualTo("act1_first_egg"));
        }

        [Test]
        public void NoEventsAtLevel0()
        {
            var state  = new PlayerStoryState();
            var engine = new StoryEngine(state);
            engine.CheckTriggers(level: 0);
            Assert.That(engine.PendingEvent, Is.Null);
        }

        [Test]
        public void OnlyOneEventPendsAtATime()
        {
            var state = new PlayerStoryState();
            // Pre-complete act1_first_egg so a second event could be eligible
            state.CompleteEvent("act1_first_egg");
            var engine = new StoryEngine(state);
            engine.CheckTriggers(level: 10);
            // Even if multiple events are eligible, only one should be pending
            Assert.That(engine.PendingEvent, Is.Not.Null);
            // The count of "pending" is implicitly 1 because PendingEvent is a single reference
        }

        [Test]
        public void ShopSecretEventTriggersAtLevel10()
        {
            var state = new PlayerStoryState();
            state.CompleteEvent("act1_first_egg");   // Prerequisite
            var engine = new StoryEngine(state);
            engine.CheckTriggers(level: 10);
            Assert.That(engine.PendingEvent, Is.Not.Null);
            Assert.That(engine.PendingEvent!.Id, Is.EqualTo("act1_shop_secret"));
        }

        [Test]
        public void ShopSecretDoesNotTriggerWithoutPrerequisite()
        {
            var state  = new PlayerStoryState();
            // act1_first_egg NOT completed
            var engine = new StoryEngine(state);
            engine.CheckTriggers(level: 10);
            // Should get act1_first_egg (still pending), not act1_shop_secret
            Assert.That(engine.PendingEvent?.Id, Is.Not.EqualTo("act1_shop_secret"));
        }

        [Test]
        public void AlreadyCompletedEventDoesNotPendAgain()
        {
            var state = new PlayerStoryState();
            state.CompleteEvent("act1_first_egg");
            var engine = new StoryEngine(state);
            engine.CheckTriggers(level: 1);
            // act1_first_egg is done — should not re-trigger
            Assert.That(engine.PendingEvent?.Id, Is.Not.EqualTo("act1_first_egg"));
        }

        // ── Event Completion Side-Effects ─────────────────────────────────────

        [Test]
        public void CompletingEventUnlocksAssociatedFeature()
        {
            var state       = new PlayerStoryState();
            var engine      = new StoryEngine(state);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;

            // Manually set the capstone event as pending
            var capstone = StoryEvent.AllEvents.First(e => e.UnlocksFeature == GameFeature.Breeding);
            engine.PendingEvent = capstone;
            engine.CompleteCurrentEvent(progression, currentLevel: 12);

            Assert.That(progression.IsFeatureUnlocked(GameFeature.Breeding), Is.True);
        }

        [Test]
        public void CompletingCapstoneAdvancesStoryAct()
        {
            var state       = new PlayerStoryState();
            var engine      = new StoryEngine(state);
            var progression = new PlayerProgression();

            var capstone = StoryEvent.AllEvents.First(e => e.AdvancesToNextAct && e.Act == StoryAct.Act1);
            engine.PendingEvent = capstone;
            engine.CompleteCurrentEvent(progression, currentLevel: 12);

            Assert.That(state.CurrentAct, Is.EqualTo(2));
        }

        [Test]
        public void CompletingEventBoostsNpcRelationship()
        {
            var state  = new PlayerStoryState();
            var engine = new StoryEngine(state);
            var prog   = new PlayerProgression();

            var firstEvent     = StoryEvent.AllEvents.First();
            int levelBefore    = state.NpcRelationshipLevels.GetValueOrDefault(firstEvent.NpcId, 0);
            engine.PendingEvent = firstEvent;
            engine.CompleteCurrentEvent(prog, currentLevel: 1);

            int levelAfter = state.NpcRelationshipLevels.GetValueOrDefault(firstEvent.NpcId, 0);
            Assert.That(levelAfter, Is.GreaterThan(levelBefore));
        }

        [Test]
        public void CompletingEventClearsPending()
        {
            var state  = new PlayerStoryState();
            var engine = new StoryEngine(state);
            var prog   = new PlayerProgression();

            engine.PendingEvent = StoryEvent.AllEvents.First();
            engine.CompleteCurrentEvent(prog, currentLevel: 1);

            // PendingEvent is either null or the NEXT chained event — not the same one
            Assert.That(engine.PendingEvent?.Id,
                Is.Not.EqualTo(StoryEvent.AllEvents.First().Id));
        }

        [Test]
        public void CompletingEventMarksIdAsCompleted()
        {
            var state       = new PlayerStoryState();
            var engine      = new StoryEngine(state);
            var prog        = new PlayerProgression();
            var firstEvent  = StoryEvent.AllEvents.First();

            engine.PendingEvent = firstEvent;
            engine.CompleteCurrentEvent(prog, currentLevel: 1);

            Assert.That(state.HasCompletedEvent(firstEvent.Id), Is.True);
        }

        // ── Save / Load Round-Trip ────────────────────────────────────────────

        [Test]
        public void SaveAndLoadStoryStateRoundTrip()
        {
            var state = new PlayerStoryState();
            state.CurrentAct = 2;
            state.CompleteEvent("act1_first_egg");
            state.CompleteEvent("act1_shop_secret");
            state.IncreaseRelationship("elder_mira", 15);

            // Flatten
            var save = new Data.StorySave
            {
                CurrentAct         = state.CurrentAct,
                CompletedEventIds  = new List<string>(state.CompletedEventIds),
                NpcRelationshipLevels   = new Dictionary<string, int>(state.NpcRelationshipLevels),
            };

            // Re-inflate
            var loaded = new PlayerStoryState
            {
                CurrentAct            = save.CurrentAct,
                CompletedEventIds     = new HashSet<string>(save.CompletedEventIds),
                NpcRelationshipLevels = new Dictionary<string, int>(save.NpcRelationshipLevels),
            };

            Assert.That(loaded.CurrentAct,                            Is.EqualTo(2));
            Assert.That(loaded.HasCompletedEvent("act1_first_egg"),   Is.True);
            Assert.That(loaded.HasCompletedEvent("act1_shop_secret"), Is.True);
            Assert.That(loaded.NpcRelationshipLevels["elder_mira"],   Is.GreaterThan(0));
        }

        [Test]
        public void SaveAndLoadPreservesAllCompletedEvents()
        {
            var state = new PlayerStoryState();
            foreach (var e in StoryEvent.AllEvents.Take(3))
                state.CompleteEvent(e.Id);

            var save = new Data.StorySave
            {
                CurrentAct        = state.CurrentAct,
                CompletedEventIds = new List<string>(state.CompletedEventIds),
                NpcRelationshipLevels  = new Dictionary<string, int>(state.NpcRelationshipLevels),
            };

            var loaded = new PlayerStoryState
            {
                CurrentAct            = save.CurrentAct,
                CompletedEventIds     = new HashSet<string>(save.CompletedEventIds),
                NpcRelationshipLevels = new Dictionary<string, int>(save.NpcRelationshipLevels),
            };

            Assert.That(loaded.CompletedEventIds.Count, Is.EqualTo(3));
        }
    }
}
